### ISSUE TITLE
shairport-sync segfault

### ISSUE THREAD

**Issue by dasl-** [mikebrady/shairport-sync#1479]
Unfortunately I'm not sure how to reproduce this. Let me know if I can do anything to help collect more data if it happens again. Here are the logs:
```
Jun  4 15:36:52 piwall10 shairport-sync:      54125.098269517 "rtsp.c:1870" Unhandled GET, path "/info?txtAirPlay&txtRAOP".
Jun  4 15:36:52 piwall10 shairport-sync:          0.000324407 "rtsp.c:1368" can't find text for response code 404. Using "Unauthorized" instead.
Jun  4 15:56:00 piwall10 shairport-sync:       1147.489752668 "rtsp.c:1870" Unhandled GET, path "/info?txtAirPlay&txtRAOP".
Jun  4 15:56:00 piwall10 shairport-sync:          0.000143834 "rtsp.c:1368" can't find text for response code 404. Using "Unauthorized" instead.
Jun  4 15:58:33 piwall10 shairport-sync:        153.764124834 "rtsp.c:2996" Connection 90. AP2 Realtime Audio Stream.
Jun  4 15:58:33 piwall10 systemd[1]: shairport-sync.service: Main process exited, code=killed, status=11/SEGV
Jun  4 15:58:33 piwall10 systemd[1]: shairport-sync.service: Failed with result 'signal'.
```

Here is my config file:
```
general =
{
  // More info about volume on pis:
  // https://github.com/dasl-/piwall2/blob/d357c3766979d473f8135448ccf36935a4fa608a/piwall2/volumecontroller.py#L22
  volume_range_db = 40; // make volume line up approximately with my own logarithmic volume algorithm on pis
  volume_max_db = 0.0; // prevent clipping on raspberry pis
  name = "piwall";
};

alsa =
{
  output_device = "hw:Headphones";
  mixer_control_name = "Headphone";
};

diagnostics =
{
  log_verbosity = 1; // "0" means no debug verbosity, "3" is most verbose.
};
```

Running in a raspberry pi 4:
```
Linux piwall10 5.10.63-v7l+ #1496 SMP Wed Dec 1 15:58:56 GMT 2021 armv7l GNU/Linux
```

Software versions:
```
% nqptp -V
Version: 1.1-dev-149-g44e74cc-dirty. Shared Memory Interface Version: 7.
% shairport-sync -V
4.1-dev-329-gead5da5a-AirPlay2-OpenSSL-Avahi-ALSA-soxr-metadata-dbus-sysconfdir:/etc
```

**Comment by mikebrady** [comment]
Thanks for the report. There is clearly some kind of intermittent bug in Shairport Sync and it would be really great to be able to reproduce it. Any progress on that would be very welcome, and of course we are trying to reproduce it here as well.

Also, I have not seen a `GET, path "/info?txtAirPlay&txtRAOP".` message before, so that's interesting. It doesn't seem to have anything to do with the crash, but it will be investigated.

Finally, thanks for the information on volume control, which I was not aware of and must have a close look at.

**Comment by dasl-** [comment]
👍 I'll let you know if I learn anything more about this bug or how to repro. I've only seen it happen once for what it's worth.

**Comment by mikebrady** [comment]
There are other references -- e.g. see #1363. I wonder if there is any connection...

**Comment by dasl-** [comment]
It happened again recently:
```
Aug 13 17:36:48 piwall10 shairport-sync:      55545.644556188 "rtsp.c:1908" Unhandled GET, path "/info?txtAirPlay&txtRAOP".
Aug 13 17:39:31 piwall10 shairport-sync:        163.428211477 "rtsp.c:1908" Unhandled GET, path "/info?txtAirPlay&txtRAOP".
Aug 13 17:40:49 piwall10 shairport-sync:         77.445608213 "rtsp.c:1908" Unhandled GET, path "/info?txtAirPlay&txtRAOP".
Aug 13 17:41:45 piwall10 shairport-sync:         56.680987096 "rtsp.c:1908" Unhandled GET, path "/info?txtAirPlay&txtRAOP".
Aug 13 17:45:36 piwall10 shairport-sync:        230.508508733 "rtsp.c:1908" Unhandled GET, path "/info?txtAirPlay&txtRAOP".
Aug 13 17:47:03 piwall10 shairport-sync:         87.462666207 "rtsp.c:1908" Unhandled GET, path "/info?txtAirPlay&txtRAOP".
Aug 13 17:47:29 piwall10 shairport-sync:         25.475380152 "rtsp.c:3058" Connection 902. AP2 Realtime Audio Stream.
Aug 13 17:47:29 piwall10 systemd[1]: shairport-sync.service: Main process exited, code=killed, status=11/SEGV
Aug 13 17:47:29 piwall10 systemd[1]: shairport-sync.service: Failed with result 'signal'.
```

```
% nqptp -V ; shairport-sync -V
Version: 1.1-dev-165-g794955b-dirty. Shared Memory Interface Version: 7.
4.1-dev-428-g9c8a3328-AirPlay2-OpenSSL-Avahi-ALSA-soxr-metadata-dbus-sysconfdir:/etc
```

I'm still unable to determine how to reproduce. I believe both segfaults occurred when my girlfriend attempted to connect to shairport-sync from her iPhone while she was playing music from the spotify iOS app.

**Comment by mikebrady** [comment]
Thanks again. I’ll keep looking. Also intrigued by those GET messages. What app / system would have been playing when those appeared? Or is it possible that Shairport Sync was not in use at the time those messages came in?

**Comment by dasl-** [comment]
I am not 100% sure, but I believe it was my girlfriend's iphone was connected to SPS playing via spotify iOS app when this bug occurred.

**Comment by dasl-** [comment]
It happened again!

SPS running on a raspberry pi 4:
```
% shairport-sync -V ; nqptp -V
4.1-dev-515-g3354f66c-AirPlay2-OpenSSL-Avahi-ALSA-soxr-metadata-dbus-sysconfdir:/etc
Version: 1.1-dev-171-ga07e9cb. Shared Memory Interface Version: 7.
```

Full log file with log level of 2 and statistics enabled (the segfault occurs all the way at the end of the logs): https://gist.github.com/dasl-/12d3361988925044f9bd3fad1f66815a

When this happened, I was connected to SPS from an intel macbook via system-wide audio output. I was streaming music from a pandora music firefox tab. I had been connected to SPS for several hours before this randomly occurred. Interestingly, there are no "Unhandled GET" messages this time.

**Comment by dasl-** [comment]
I've enabled coredumps on my system by making the following changes:

1) I created a file `/etc/sysctl.d/core.conf` with the contents:
```
kernel.core_pattern = /tmp/coredump-%e-sig%s-user%u-group%g-pid%p-time%t
```
A restart is required after this step.

2) I've modified the shairport-sync systemd service file to set `LimitCORE=infinity` in the `[Service]` stanza.

So hopefully we'll have more data the next time this happens.

**Comment by mikebrady** [comment]
It would be great alright. Also I'm trying to figure out how to attach `gdb` automagically when a `SEGENV` occurs. Apparently it is possible. Then I could get it to do a stack trace on all the threads, maybe see what has happened.

A good thing to do is to turn off compiler optimisation, to get a slightly better chance of understanding what comes out. What I do at the configuration stage is to prefix it as follows:

```
CFLAGS="-O0 -g" CXXFLAGS="-O0 -g" ./configure ...
```

**Comment by dasl-** [comment]
Ok, here's what I've done: https://github.com/dasl-/pitools/commit/c987c537e0e536f1a4e26fec510b7ee3ce9f9536

Let me know if that looks right. I've not worked with C or coredumps much before, so I could easily be doing something wrong!

**Comment by mikebrady** [comment]
> Let me know if that looks right.

I’m afraid I don’t have any guidance or offer you — I haven’t done this either 😬.

**Comment by mikebrady** [comment]
Hi there. We found a Spotify-specific bug that caused Shairport Sync to crash -- see #1551. Specifically, Spotify occasionally omits an important decryption parameter. Now, instead of crashing, Shairport Sync simply stays silent. I wonder if that is the case of the crates you experienced.

**Comment by dasl-** [comment]
Interesting. I'm not sure if I was using spotify in every instance of this segfault, but it's a possibility. Unfortunately I've still not had any segfaults since I enabled coredumps 🥲 

I'll update to the latest development branch code regardless.

**Comment by dasl-** [comment]
The logs I got for these segfaults (posted in an earlier comment: https://github.com/mikebrady/shairport-sync/issues/1479#issuecomment-1230476995 ) are very similar to the logs I got in another issue I filed about a segfault ( https://github.com/mikebrady/shairport-sync/issues/1571 ). Furthermore, that issue includes a coredump. So I'm going to close this issue since the other issue now has more information (a coredump). Potentially both of these issues I opened have the same root cause.
